<a href="https://colab.research.google.com/github/Lotfullah21/PyLibraries/blob/main/Pandas_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
df = pd.DataFrame({"name":[
                           "Lotfullah Andishmand",
                           "M.D Asif Sharifi",
                           "Ayaan Ali",
                           "Yesha Sweni"
],
"Age":[24,43,21,22],
"Sex":['male','male','male','female']
})
df

name  Age     Sex
0  Lotfullah Andishmand   24    male
1      M.D Asif Sharifi   43    male
2             Ayaan Ali   21    male
3           Yesha Sweni   22  female

In [3]:
age = pd.Series([32,1,21,12])
# columns are considered as series
age.max()
age.describe()

count     4.000000
mean     16.500000
std      13.178265
min       1.000000
25%       9.250000
50%      16.500000
75%      23.750000
max      32.000000
dtype: float64

In [4]:
# https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%202/data/FuelConsumptionCo2.csv
df = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%202/data/FuelConsumptionCo2.csv')
df.head()

MODELYEAR   MAKE       MODEL VEHICLECLASS  ENGINESIZE  CYLINDERS  \
0       2014  ACURA         ILX      COMPACT         2.0          4   
1       2014  ACURA         ILX      COMPACT         2.4          4   
2       2014  ACURA  ILX HYBRID      COMPACT         1.5          4   
3       2014  ACURA     MDX 4WD  SUV - SMALL         3.5          6   
4       2014  ACURA     RDX AWD  SUV - SMALL         3.5          6   

  TRANSMISSION FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  \
0          AS5        Z                   9.9                  6.7   
1           M6        Z                  11.2                  7.7   
2          AV7        Z                   6.0                  5.8   
3          AS6        Z                  12.7                  9.1   
4          AS6        Z                  12.1                  8.7   

   FUELCONSUMPTION_COMB  FUELCONSUMPTION_COMB_MPG  CO2EMISSIONS  
0                   8.5                        33           196  
1                   9.6                        29           221  
2                   5.9                        48           136  
3                  11.1                        25           255  
4                  10.6                        27           244

# Or
we can directly put the name of the file


In [5]:
!wget -O FuelConsumption.csv https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%202/data/FuelConsumptionCo2.csv

--2022-05-26 13:23:58--  https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%202/data/FuelConsumptionCo2.csv
Resolving cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud (cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud)... 169.63.118.104
Connecting to cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud (cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud)|169.63.118.104|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 72629 (71K) [text/csv]
Saving to: ‘FuelConsumption.csv’

FuelConsumption.csv 100%[===================>]  70.93K  --.-KB/s    in 0.04s   

2022-05-26 13:23:58 (1.85 MB/s) - ‘FuelConsumption.csv’ saved [72629/72629]



In [6]:
df = pd.read_csv("FuelConsumption.csv")
df.head()

MODELYEAR   MAKE       MODEL VEHICLECLASS  ENGINESIZE  CYLINDERS  \
0       2014  ACURA         ILX      COMPACT         2.0          4   
1       2014  ACURA         ILX      COMPACT         2.4          4   
2       2014  ACURA  ILX HYBRID      COMPACT         1.5          4   
3       2014  ACURA     MDX 4WD  SUV - SMALL         3.5          6   
4       2014  ACURA     RDX AWD  SUV - SMALL         3.5          6   

  TRANSMISSION FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  \
0          AS5        Z                   9.9                  6.7   
1           M6        Z                  11.2                  7.7   
2          AV7        Z                   6.0                  5.8   
3          AS6        Z                  12.7                  9.1   
4          AS6        Z                  12.1                  8.7   

   FUELCONSUMPTION_COMB  FUELCONSUMPTION_COMB_MPG  CO2EMISSIONS  
0                   8.5                        33           196  
1                   9.6                        29           221  
2                   5.9                        48           136  
3                  11.1                        25           255  
4                  10.6                        27           244

In [7]:
# change the data to an excel file
df.to_excel("FuelConsumption.xlsx", sheet_name="passengers", index=False)
df.head(4)

MODELYEAR   MAKE       MODEL VEHICLECLASS  ENGINESIZE  CYLINDERS  \
0       2014  ACURA         ILX      COMPACT         2.0          4   
1       2014  ACURA         ILX      COMPACT         2.4          4   
2       2014  ACURA  ILX HYBRID      COMPACT         1.5          4   
3       2014  ACURA     MDX 4WD  SUV - SMALL         3.5          6   

  TRANSMISSION FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  \
0          AS5        Z                   9.9                  6.7   
1           M6        Z                  11.2                  7.7   
2          AV7        Z                   6.0                  5.8   
3          AS6        Z                  12.7                  9.1   

   FUELCONSUMPTION_COMB  FUELCONSUMPTION_COMB_MPG  CO2EMISSIONS  
0                   8.5                        33           196  
1                   9.6                        29           221  
2                   5.9                        48           136  
3                  11.1                        25           255

In [8]:
df.dtypes

MODELYEAR                     int64
MAKE                         object
MODEL                        object
VEHICLECLASS                 object
ENGINESIZE                  float64
CYLINDERS                     int64
TRANSMISSION                 object
FUELTYPE                     object
FUELCONSUMPTION_CITY        float64
FUELCONSUMPTION_HWY         float64
FUELCONSUMPTION_COMB        float64
FUELCONSUMPTION_COMB_MPG      int64
CO2EMISSIONS                  int64
dtype: object

In [9]:
titanic = pd.read_excel("FuelConsumption.xlsx", sheet_name="passengers")
titanic.head()

MODELYEAR   MAKE       MODEL VEHICLECLASS  ENGINESIZE  CYLINDERS  \
0       2014  ACURA         ILX      COMPACT         2.0          4   
1       2014  ACURA         ILX      COMPACT         2.4          4   
2       2014  ACURA  ILX HYBRID      COMPACT         1.5          4   
3       2014  ACURA     MDX 4WD  SUV - SMALL         3.5          6   
4       2014  ACURA     RDX AWD  SUV - SMALL         3.5          6   

  TRANSMISSION FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  \
0          AS5        Z                   9.9                  6.7   
1           M6        Z                  11.2                  7.7   
2          AV7        Z                   6.0                  5.8   
3          AS6        Z                  12.7                  9.1   
4          AS6        Z                  12.1                  8.7   

   FUELCONSUMPTION_COMB  FUELCONSUMPTION_COMB_MPG  CO2EMISSIONS  
0                   8.5                        33           196  
1                   9.6                        29           221  
2                   5.9                        48           136  
3                  11.1                        25           255  
4                  10.6                        27           244

In [10]:
model = df['MODEL']
model.head()


0           ILX
1           ILX
2    ILX HYBRID
3       MDX 4WD
4       RDX AWD
Name: MODEL, dtype: object

In [11]:
Emission = df['CO2EMISSIONS']
Emission.head()
df['CO2EMISSIONS'].shape


(1067,)

In [12]:
# The inner square brackets define a Python list with column names, whereas the outer brackets
#  are used to select the data from a pandas DataFrame as seen in the previous example.
Co2_Model = df[['CO2EMISSIONS','MODEL','ENGINESIZE']]
Co2_Model.head()


CO2EMISSIONS       MODEL  ENGINESIZE
0           196         ILX         2.0
1           221         ILX         2.4
2           136  ILX HYBRID         1.5
3           255     MDX 4WD         3.5
4           244     RDX AWD         3.5

In [13]:
Co2_Model.head()

CO2EMISSIONS       MODEL  ENGINESIZE
0           196         ILX         2.0
1           221         ILX         2.4
2           136  ILX HYBRID         1.5
3           255     MDX 4WD         3.5
4           244     RDX AWD         3.5

In [14]:
Co2_Model.shape


(1067, 3)

# Filtering

In [15]:
#  filtering the engine size < 2
E_greater_than2  = df[df["ENGINESIZE"]>2]
E_greater_than2.head()

MODELYEAR   MAKE    MODEL VEHICLECLASS  ENGINESIZE  CYLINDERS TRANSMISSION  \
1       2014  ACURA      ILX      COMPACT         2.4          4           M6   
3       2014  ACURA  MDX 4WD  SUV - SMALL         3.5          6          AS6   
4       2014  ACURA  RDX AWD  SUV - SMALL         3.5          6          AS6   
5       2014  ACURA      RLX     MID-SIZE         3.5          6          AS6   
6       2014  ACURA       TL     MID-SIZE         3.5          6          AS6   

  FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  FUELCONSUMPTION_COMB  \
1        Z                  11.2                  7.7                   9.6   
3        Z                  12.7                  9.1                  11.1   
4        Z                  12.1                  8.7                  10.6   
5        Z                  11.9                  7.7                  10.0   
6        Z                  11.8                  8.1                  10.1   

   FUELCONSUMPTION_COMB_MPG  CO2EMISSIONS  
1                        29           221  
3                        25           255  
4                        27           244  
5                        28           230  
6                        28           232

In [31]:
# Cecking for the rows which has EngineSize greater than 2
df['ENGINESIZE']>2
above_4  = df[['CYLINDERS','ENGINESIZE']]>4
above_4.shape
above_4

CYLINDERS  ENGINESIZE
0         False       False
1         False       False
2         False       False
3          True       False
4          True       False
...         ...         ...
1062       True       False
1063       True       False
1064       True       False
1065       True       False
1066       True       False

[1067 rows x 2 columns]

In [35]:
# we want to see if any egnine size row is null or not
row_not_na = df['ENGINESIZE'].notna()
row_not_na
row_not_na.head()

0    True
1    True
2    True
3    True
4    True
Name: ENGINESIZE, dtype: bool

#### we are interested in cars that their engine size is greater than 4.


In [39]:
Car_model = df.loc[df['ENGINESIZE']>4,'MODEL']
Car_model.head()

12             DB9
13          RAPIDE
14      V8 VANTAGE
15      V8 VANTAGE
16    V8 VANTAGE S
Name: MODEL, dtype: object

### We are interested in rows 32 till 50 and columns 5 to 9

In [54]:
df.iloc[32:50,0:9]


MODELYEAR  MAKE                 MODEL           VEHICLECLASS  ENGINESIZE  \
32       2014  AUDI   A8 TDI CLEAN DIESEL               MID-SIZE         3.0   
33       2014  AUDI                   A8L              FULL-SIZE         3.0   
34       2014  AUDI                   A8L              FULL-SIZE         4.0   
35       2014  AUDI                   A8L              FULL-SIZE         6.3   
36       2014  AUDI  A8L TDI CLEAN DIESEL              FULL-SIZE         3.0   
37       2014  AUDI       ALLROAD QUATTRO  STATION WAGON - SMALL         2.0   
38       2014  AUDI                    Q5            SUV - SMALL         2.0   
39       2014  AUDI                    Q5            SUV - SMALL         3.0   
40       2014  AUDI             Q5 HYBRID            SUV - SMALL         2.0   
41       2014  AUDI   Q5 TDI CLEAN DIESEL            SUV - SMALL         3.0   
42       2014  AUDI                    Q7         SUV - STANDARD         3.0   
43       2014  AUDI   Q7 TDI CLEAN DIESEL         SUV - STANDARD         3.0   
44       2014  AUDI                    R8             TWO-SEATER         4.2   
45       2014  AUDI                    R8             TWO-SEATER         4.2   
46       2014  AUDI                    R8             TWO-SEATER         5.2   
47       2014  AUDI                    R8             TWO-SEATER         5.2   
48       2014  AUDI             R8 SPYDER             TWO-SEATER         4.2   
49       2014  AUDI             R8 SPYDER             TWO-SEATER         4.2   

    CYLINDERS TRANSMISSION FUELTYPE  FUELCONSUMPTION_CITY  
32          6          AS8        D                  10.0  
33          6          AS8        Z                  13.1  
34          8          AS8        Z                  14.7  
35         12          AS8        Z                  18.2  
36          6          AS8        D                  10.1  
37          4          AS8        Z                  11.8  
38          4          AS8        Z                  12.0  
39          6          AS8        Z                  12.9  
40          4          AS8        Z                   9.9  
41          6          AS8        D                  10.3  
42          6          AS8        Z                  15.1  
43          6          AS8        D                  12.9  
44          8           A7        Z                  17.6  
45          8           M6        Z                  21.2  
46         10           A7        Z                  18.8  
47         10           M6        Z                  21.1  
48          8           A7        Z                  17.6  
49          8           M6        Z                  21.2

In [57]:
df.iloc[0:5,4] = 'Tesla'
df.head()

MODELYEAR   MAKE  MODEL VEHICLECLASS ENGINESIZE  CYLINDERS TRANSMISSION  \
0       2014  Tesla  Tesla        Tesla      Tesla          4          AS5   
1       2014  Tesla  Tesla        Tesla      Tesla          4           M6   
2       2014  Tesla  Tesla        Tesla      Tesla          4          AV7   
3       2014  Tesla  Tesla        Tesla      Tesla          6          AS6   
4       2014  Tesla  Tesla        Tesla      Tesla          6          AS6   

  FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  FUELCONSUMPTION_COMB  \
0        Z                   9.9                  6.7                   8.5   
1        Z                  11.2                  7.7                   9.6   
2        Z                   6.0                  5.8                   5.9   
3        Z                  12.7                  9.1                  11.1   
4        Z                  12.1                  8.7                  10.6   

   FUELCONSUMPTION_COMB_MPG  CO2EMISSIONS  
0                        33           196  
1                        29           221  
2                        48           136  
3                        25           255  
4                        27           244

### We can assign names to a specific columns or rows also,for example here we are going to assign tesla for the first 5 elemenst of clumn 1 which is the model type


In [53]:
df

MODELYEAR   MAKE     MODEL    VEHICLECLASS  ENGINESIZE  CYLINDERS  \
0          2014  Tesla     Tesla           Tesla         2.0          4   
1          2014  Tesla     Tesla           Tesla         2.4          4   
2          2014  Tesla     Tesla           Tesla         1.5          4   
3          2014  Tesla     Tesla           Tesla         3.5          6   
4          2014  Tesla     Tesla           Tesla         3.5          6   
...         ...    ...       ...             ...         ...        ...   
1062       2014  VOLVO  XC60 AWD     SUV - SMALL         3.0          6   
1063       2014  VOLVO  XC60 AWD     SUV - SMALL         3.2          6   
1064       2014  VOLVO  XC70 AWD     SUV - SMALL         3.0          6   
1065       2014  VOLVO  XC70 AWD     SUV - SMALL         3.2          6   
1066       2014  VOLVO  XC90 AWD  SUV - STANDARD         3.2          6   

     TRANSMISSION FUELTYPE  FUELCONSUMPTION_CITY  FUELCONSUMPTION_HWY  \
0             AS5        Z                   9.9                  6.7   
1              M6        Z                  11.2                  7.7   
2             AV7        Z                   6.0                  5.8   
3             AS6        Z                  12.7                  9.1   
4             AS6        Z                  12.1                  8.7   
...           ...      ...                   ...                  ...   
1062          AS6        X                  13.4                  9.8   
1063          AS6        X                  13.2                  9.5   
1064          AS6        X                  13.4                  9.8   
1065          AS6        X                  12.9                  9.3   
1066          AS6        X                  14.9                 10.2   

      FUELCONSUMPTION_COMB  FUELCONSUMPTION_COMB_MPG  CO2EMISSIONS  
0                      8.5                        33           196  
1                      9.6                        29           221  
2                      5.9                        48           136  
3                     11.1                        25           255  
4                     10.6                        27           244  
...                    ...                       ...           ...  
1062                  11.8                        24           271  
1063                  11.5                        25           264  
1064                  11.8                        24           271  
1065                  11.3                        25           260  
1066                  12.8                        22           294  

[1067 rows x 13 columns]

# Technical summary

<!-- for  -->

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067 entries, 0 to 1066
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   MODELYEAR                 1067 non-null   int64  
 1   MAKE                      1067 non-null   object 
 2   MODEL                     1067 non-null   object 
 3   VEHICLECLASS              1067 non-null   object 
 4   ENGINESIZE                1067 non-null   float64
 5   CYLINDERS                 1067 non-null   int64  
 6   TRANSMISSION              1067 non-null   object 
 7   FUELTYPE                  1067 non-null   object 
 8   FUELCONSUMPTION_CITY      1067 non-null   float64
 9   FUELCONSUMPTION_HWY       1067 non-null   float64
 10  FUELCONSUMPTION_COMB      1067 non-null   float64
 11  FUELCONSUMPTION_COMB_MPG  1067 non-null   int64  
 12  CO2EMISSIONS              1067 non-null   int64  
dtypes: float64(4), int64(4), object(5)
memory usage: 108.5+ KB


We have 1067 rows, 12 columns, we non_null shows that we do not have any missing value, we can see for example the **MODELYEAR** values are in the form of integers and **FUELTYPE** datatypes are in the form of object.
we have different datatypes. at the end we can see how much memory we have used for this dataset

<!-- -->